File: BOJ 2/11266.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
#include <cstring>
using namespace std;
int n, m,t;
vector<vector<int>> a;
int dis[10001];
bool cutvertex[10001];
vector<int> v;
int dfs(int here,bool ok) {
	dis[here] = ++t;
	int child = 0;
	int ret = dis[here];
	for (int i = 0; i < a[here].size(); i++) {
		int next = a[here][i];
		if (!dis[next]) {
			child++;
			int r = dfs(next, false);
			if (!ok && r >= dis[here]) {
				cutvertex[here] = true;
			}
			ret = min(ret, r);
		}
		else ret = min(ret, dis[next]);
	}
	if (ok && child >= 2) cutvertex[here] = true;
	return ret;
}
int main() {
	scanf("%d%d", &n, &m);
	a.resize(n + 1);
	for (int i = 0; i < m; i++) {
		int first, second;
		scanf("%d%d", &first, &second);
		a[first].push_back(second);
		a[second].push_back(first);
	}

	for (int i = 1; i <= n; i++) {
		if (!dis[i]) dfs(i, true);
	}
	for (int i = 1; i <= n; i++) {
		if (cutvertex[i]) v.push_back(i);
	}
	printf("%d\n", v.size());
	for (int i = 0; i < v.size(); i++) printf("%d ", v[i]);
	return 0;
}

File: BOJ(Java 공부용)/11266.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

import static java.lang.Integer.min;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int t = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);
        List<Integer>[] graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        int[] dis = new int[n + 1];
        boolean[] cut = new boolean[n + 1];
        ArrayList<Integer> v = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int here = Integer.parseInt(st.nextToken()), next = Integer.parseInt(st.nextToken());
            graph[here].add(next);
            graph[next].add(here);
        }
        for (int i = 1; i <= n; i++) {
            if (dis[i] == 0) dfs(i, true, cut, dis, graph);
        }
        for (int i = 1; i <= n; i++) {
            if (cut[i]) v.add(i);
        }
        bw.write(v.size() + "\n");
        for (Integer e : v) {
            bw.write(e + " ");
        }
        bw.flush();
        bw.close();
    }

    private static int dfs(int here, boolean ok, boolean[] cut, int[] dis, List<Integer>[] graph) {
        dis[here] = ++t;
        int child = 0;
        int ret = dis[here];
        for (int i = 0; i < graph[here].size(); i++) {
            int next = graph[here].get(i);
            if (dis[next] == 0) {
                child++;
                int r = dfs(next, false, cut, dis, graph);
                if (!ok && r >= dis[here]) {
                    cut[here] = true;
                }
                ret = min(ret, r);
            } else ret = min(ret, dis[next]);
        }
        if (ok && child >= 2) cut[here] = true;
        return ret;
    }
}